doc/subsystems/codegen/modules.rst

```diff
@@ -0,0 +1,316 @@
+                     declared by device node.
+                     If the 'interrupts-names' property is provided in the node,
+                     isr names will be generated using matching values as
+                     function postfixes.
+    :param init_func: Name of the driver init function (e.g. 'i2c_stm32_init').
+    :param api: Name of the driver api structure  (e.g. 'api_funcs').
+    :param data_struct: Two elements python list providing elements for driver
+                        '_data' structure generation.
+    :param config_struct: Two elements python list providing elements for driver
+                          '_config' structure generation.
+
+'data_struct' and 'config_struct' will be processed the same way:
+
+* First element (e.g. 'i2c_stm32_config') should be the structure name.
+* Second element is a 'c' template code between triple double quotes (""" """).
+    It should provide the expected code to be generated for the structure.
+    For instance:
+
+.. code-block:: python
+
+    """
+    	.i2c = (I2C_TypeDef *)${reg/0/address/0},
+    	.pclken = {
+    			.enr = ${clocks/0/bits},
+    			.bus = ${clocks/0/bus},
+    	},
+    #ifdef CONFIG_I2C_STM32_INTERRUPT
+    	.irq_config_func = st_stm32_i2c_v1_${node_index}_config_irq,
+    #endif
+    	.bitrate = ${clock-frequency},
+    """
+
+If the second element of 'data_struct' or 'config_struct' list is not provided,
+an empty structure is generated.
+
+Finally, for the above depicted example, 'device_declare' will generate,
+for device instance 'I2C1':
+
+.. code-block:: c
+
+    #ifdef CONFIG_I2C_STM32_INTERRUPT
+    DEVICE_DECLARE(st_stm32_i2c_v1_i2c_1);
+    static void st_stm32_i2c_v1_i2c_1_config_irq(struct device *dev)
+    {
+    	IRQ_CONNECT(31,
+    		0,
+    		stm32_i2c_isr_event,
+    		DEVICE_GET(st_stm32_i2c_v1_i2c_1),
+    		0);
+    	irq_enable(31);
+    	IRQ_CONNECT(32,
+    		0,
+    		stm32_i2c_isr_error,
+    		DEVICE_GET(st_stm32_i2c_v1_i2c_1),
+    		0);
+    	irq_enable(32);
+    }
+    #endif /* CONFIG_I2C_STM32_INTERRUPT */
+
+    static const struct i2c_stm32_config st_stm32_i2c_v1_i2c_1_config = {
+    	.i2c = (I2C_TypeDef *)0x40005400,
+    	.pclken = {
+    		.enr = 131072,
+    		.bus = 2,
+    	},
+    #ifdef CONFIG_I2C_STM32_INTERRUPT
+    	.irq_config_func = st_stm32_i2c_v1_i2c_1_config_irq,
+    #endif
+    	.bitrate = 400000,
+    };
+
+    static struct i2c_stm32_data st_stm32_i2c_v1_i2c_1_data = {};
+
+    DEVICE_AND_API_INIT(st_stm32_i2c_v1_i2c_1,
+    	"I2C_1",
+    	i2c_stm32_init,
+    	&st_stm32_i2c_v1_i2c_1_data,
+    	&st_stm32_i2c_v1_i2c_1_config,
+    	POST_KERNEL,
+    	CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+    	&api_funcs);
+
+Raw declaration of a single device instance
+-------------------------------------------
+
+.. function:: devicedeclare.device_declare_single(device_config, driver_name, device_init, device_levels device_prio, device_api, device_info)
+
+    Generate device instances code for a device instance that:
```

Comment:
As I understand this is the low level API you propose that should be the base for other APIs.
Let me comment on this part rather than code actual code for now.